### PR TITLE
Add insert page feature

### PR DESCRIPTION
**Pull request by andreehultgren**
## Insert page feature 
This feature allows you to insert a blank page at page X in the document. I needed this for my own project, and wanted to add it to the repo. If this is wanted in the repo, I can add the unit testing and documentation as well.

**Checklist**:

- [ ] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [ ] A unit test is covering the code added / modified by this PR

- [ ] This PR is ready to be merged 

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [ ] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/PyFPDF/fpdf2/blob/master/LICENSE).

### CODE REVIEW

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -751,6 +751,46 @@ def alias_nb_pages(self, alias="{nb}"):
...
+                Can be configured globally through the `.page_duration` FPDF property.
+                As of june 2021, onored by Adobe Acrobat reader, but ignored by Sumatra PDF reader.
+            transition (Transition child class): optional visual transition to use when moving
+                from another page to the given page during a presentation.
+                Can be configured globally through the `.page_transition` FPDF property.
+                As of june 2021, onored by Adobe Acrobat reader, but ignored by Sumatra PDF reader.
+        """
+        # Add a page at the end of the pdf
+        self.page = len(self.pages)
+        self.add_page(orientation, format, same, duration, transition)
```
Comment (Lucas-C) [inline fpdf/fpdf.py:782]: Unless `insert_location == self.pages_count`, the document last page should stay the same,
and hence its value for `orientation`, `format`, `duration` & `transition` should be taken from the last page before calling this method.

The same reasoning applies to all pages after `inert_location`: this method will break all those per-page settings by shfting them to previous page

**Review by Lucas-C — COMMENTED** [review]
Sadly, this approach won't work as it is.
There are several references to `self.page` that would be broken there:
* in [add_link()](https://github.com/PyFPDF/fpdf2/blob/master/fpdf/fpdf.py#L1984)
* outline sections: https://github.com/PyFPDF/fpdf2/blob/master/fpdf/fpdf.py#L4385
* marked content IDs: https://github.com/PyFPDF/fpdf2/blob/master/fpdf/fpdf.py#L3805
* quad points: https://github.com/PyFPDF/fpdf2/blob/master/fpdf/fpdf.py#L3071

All of those `.page` references would have to be updated.

Why do you need this method exactly?
What prevents you from rendering the PDF document in order?

File: fpdf/fpdf.py (hunk 1 above)
Comment (andreehultgren) [reply fpdf/fpdf.py:782]: Line 781 ensures that we are at the last page.
```python
self.page = len(self.pages)
```

**Comment by andreehultgren** [comment]
The PR is only a suggestion of a change that can be made. I was able to solve the same thing with the following steps with better reliability:

 1. Add document in dry-mode (it allocates the space, but doesn't add anything in the document)
 2. Write the rest of the document
 3. Go back to the dry area and add the actual content

This is useful for creating a table of content with links to pages in the document. I know there is a function for TOC, but when I use the built-in, the TOC isn't placed in the start of the document.

**Comment by Lucas-C** [comment]
> This is useful for creating a table of content with links to pages in the document. I know there is a function for TOC, but when I use the built-in, the TOC isn't placed in the start of the document.

I'm a bit sad you have to craft your own solution because [the existing `insert_toc_placeholder()` solution](https://pyfpdf.github.io/fpdf2/DocumentOutlineAndTableOfContents.html) does not work for you 😢 

Could we maybe try to improve `insert_toc_placeholder()` so that it works for you?
Our unit tests manage to insert a ToC on the first page, so I'm not sure of the origin of your problem:
https://github.com/PyFPDF/fpdf2/blob/master/test/outline/test_outline.py#L13

**Comment by andreehultgren** [comment]
The issue with `insert_toc_placeholder` is that you need to know the amount of pages before you render the document. 
If the guess is incorrect, then the allocated space can be too much or too little.

An alternative to my PR is to improve the function `_insert_table_of_contents` such that it injects all the pages neccesary where the user places a `insert_toc_placeholder` or a renamed alternative `insert_toc_flag`.

So the workflow could be:
1. Generate a cover page
2. Place a flag for TOC
3. Generate document
4. Insert TOC at flag

**Comment by Lucas-C** [comment]
OK, so your problem is related to #136

I agree that requiring to guess the number of pages of the ToC is annoying limitation.

However, I do not understand how your suggest 4-steps workflow will solve this problem.
Your "flag" seems to have the exact same role as `insert_toc_placeholder()`:
https://pyfpdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.insert_toc_placeholder

How would your flag-based solution behave differently?

**Comment by andreehultgren** [comment]
The flag does not behave anny differently, they are the same.

The fix would be to remove the ToC limitation, so yes. It would be the same as the issue you references. I will close the PR and see if I can solve the issue - as it is up for grabs.

**Pull request closed without merging**